client.go

```diff
@@ -64,57 +67,46 @@ type Request struct {
 }
 
 type idPool struct {
-	IDs chan uint16
+	IDs uint16
+
+	Used *sync.Map
+	Lock *sync.Mutex
 }
 
 // AllocID implements Client.AllocID
 func (p *idPool) Alloc() uint16 {
-	return <-p.IDs
-}
-
-// ReleaseID implements Client.ReleaseID
-func (p *idPool) Release(id uint16) {
-	go func() {
-		// release the ID back to channel for reuse
-		// use goroutine to prev0, ent blocking ReleaseID
-		p.IDs <- id
-	}()
-}
-
-func newIDs(limit uint32) (p idPool) {
+	p.Lock.Lock()
+next:
+	idx := p.IDs
+	if idx == MaxUint {
+		// reset
+		idx = 1
+		p.IDs = 1
+	}
+	p.IDs++
 
-	// sanatize limit
-	if limit == 0 || limit > 65535 {
-		// Note: limit is the size of the pool
-		// Since 0 cannot be requestId, the effective
-		// pool is from 1 to 65535, hence size is 65535.
-		limit = 65535
+	if _, inuse := p.Used.Load(idx); inuse {
+		// Allow other go-routine to take priority
+		// to prevent spinlock here
+		runtime.Gosched()
```

Comment:
This context switch can solve the previous test where only 2 goroutines are in the picture. But can it solve the case where `pool.Alloc()` is called on multiple goroutines before `pool.Release()` is called? Please update the blocking test to demonstrate. Thanks.

Reply:
I'll have a look into extending the test.

Reply:
Theoretical answer: Yes, because of the lock on Alloc.